File: seminar_3/task_4/task4.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

template <typename T>
ostream& operator<<(ostream &fout, const vector <T> &v){
    for(auto curr : v)
        fout<<curr<<" ";
    return fout;
}

template <typename T>
ostream& operator<<(ostream &fout, const unordered_map <T, T> &v){
    for(auto curr : v)
        fout<<curr.first<<": "<<curr.second<<endl;
    return fout;
}

vector <int> two_sum(const vector <int> &target, int data){
    unordered_map<int, int> cash;
    for(int i=0; i<int(target.size()); i++){
        if(cash.find(target[i]) == cash.end())
            cash[data-target[i]] = i;
        else
            return vector<int>{cash[target[i]], i};
    }
    return vector<int>{};
}

int main(void){
    cout<<two_sum({1, 3, 5, 6, 4, 8}, 12)<<endl;
    cout<<two_sum({-13, 56, 36, 9, -4, 7}, 23);
    return 0;
}

File: seminar_3/task_4/task_4.py
class HashTable:
    def __init__(self):
        self.table = {}

    def contains(self, key):
        return key in self.table

    def insert(self, key, value):
        self.table[key] = value

    def get(self, key):
        return self.table.get(key)


def twoSum(data, target):
    ht = HashTable()
    for i in range(len(data)):
        if ht.contains(data[i]): # среди ключей нашей таблицы ищем data[i]
            return [i, ht.table[data[i]]]
        diff = target - data[i]
        ht.table[diff] = i
    return []


""" a = [6, 4, 3, 2, 8, 5]
target = 11
print(twoSum(a, target)) """